src/gif.rs

```diff
@@ -53,7 +54,14 @@ pub struct GifDecoder<R: Read> {
 impl<R: Read> GifDecoder<R> {
     /// Creates a new decoder that decodes the input steam ```r```
     pub fn new(r: R) -> ImageResult<GifDecoder<R>> {
+        Self::new_with_limits(r, DecodingLimits::default())
+    }
+
+    /// Same as `new`, but allows you to set limits.
+    pub fn new_with_limits(r: R, limits: DecodingLimits) -> ImageResult<GifDecoder<R>> {
+        let gif_limits = gif::MemoryLimit(min(u32::max_value() as usize, limits.buffer_limit) as u32);
         let mut decoder = gif::Decoder::new(r);
+        decoder.set(gif_limits);
```

Comment:
Here we have `new_with_limits`, in the case of `io::Reader` there is `set_limits`. These should agree on one kind of interface.

Reply:
I have used `new_with_limits` when creating the decoders in order to set the limits as early as possible when starting decoding, most of the decoders start reading data in their `new` functions and I thought it was a good idea to set the limits before this.

Reply:
So because decoders can't delay reading metadata they have to do this in fallible constructors. This is apparently good enough so we don't have to modify it. I'm mostly wondering if the reasoning to make the metadata queries non-fallible in 2957fc25ba798fe42a661dccd9544c3e2a9a3167 still holds true.

cc: @fintelia

Reply:
Are there any formats were the resource use from decoding metadata is a concern? It generally just amounts to reading constant size headers, right?

The motivation for requiring metadata decoding in the constructor was that it streamlined the rest of the interface. If the constructor was infallible instead, then implementors would have to do lazy loading of metadata and any users would have to deal with possible I/O errors on every metadata call. Any failure to retrieve the basic image metadata necessarily means that loading the rest of the image cannot succeed.

And once you have the image metadata, a lot of the limits logic can actually live in the caller. Want to make sure you don't end up with an image that is too large? Just check the dimensions before calling decode. Intermediate buffers are still a concern, but long term I think we could have decoders for all the formats that use only a constant amount of additional space.

Reply:
> Are there any formats were the resource use from decoding metadata is a concern? It generally just amounts to reading constant size headers, right?

I can't think of an immediate concern, we don't track constant space usage in any case since that would include the stack, local variables, etc. which is infeasible and not as important. I was just curious. But it is not entirely clear to me if *all* formats have a constant resource usage for reading metadata. They have `io::Seek` but need to refrain from reading any variable length metadata (EXIF comments, etc.) in the constructor.